### ISSUE TITLE
Calculating isosurfaces

### ISSUE THREAD

**Issue by hetland** [xgcm/xgcm#144]
I just left [a comment in another package](https://github.com/lesommer/oocgcm/issues/42) about an xarray native isoslice routine. Would it be useful to include this in xgcm?

One one hand, it's not specifically part of xgcm's core mission, but on the other hand, it seems related. There is a coordinate hack, but these coordinates are eventually lost in the result. The algorithm requires monotonic data, like [ciso](https://github.com/ioos/ciso). It is otherwise pure xarray, so can be lazily evaluated. The function is (edited 4/7/2020 to fix a typo in the return statement):

```
def xisoslice(ds, coord, iso_array, iso_value, projected_array):
    '''Calculate an isosurface
    
    This function calculates the value of projected_array on 
    an isosurface in the array iso_array defined by iso_val.
    
    Inputs:
    ds:              xarray dataset: contains dataarrays iso_array and projected_array
    coord:           string: coordinate associated with the dimension along which to project
    iso_array:       string: array in ds
    iso_value:       float:  value of the isosurface in iso_array
    projected_array: string: array in ds on which to project values onto the isosurface
    
    Output:
    iso_values:      xarray array: values of projected_array on the isosurface   
    '''
    Nm = len(ds[coord]) - 1

    lslice = {coord: slice(None, -1)}
    uslice = {coord: slice(1, None)}

    prop = ds[iso_array] - iso_value

    propl = prop.isel(**lslice)
    propl.coords[coord] = np.arange(Nm)
    propu = prop.isel(**uslice)
    propu.coords[coord] = np.arange(Nm)

    zc = xr.where((propu*propl)<0.0, 1.0, 0.0)

    varl = ds[projected_array].isel(**lslice)
    varl.coords[coord] = np.arange(Nm)
    varu = ds[projected_array].isel(**uslice)
    varu.coords[coord] = np.arange(Nm)

    propl = (propl*zc).sum(coord)
    propu = (propu*zc).sum(coord)
    varl = (varl*zc).sum(coord)
    varu = (varu*zc).sum(coord)

    return varl - propl*(varu-varl)/(propu-propl)
```

Use cases:

Getting salinity at a specified depth:
```
result = xisoslice(ds, 's_rho', 'z_rho', -20, 'salt')
```

Getting the depth of a salinity isosurface:
```
result = xisoslice(ds, 's_rho', 'salt', 28.0, 'z_rho')
```

Getting a cross-section along a constant longitude:
```
salt_slice = xisoslice(ds, 'xi_rho', 'lon_rho', -92.0, 'salt')
salt_slice.coords['z_rho'] = xisoslice(ds, 'xi_rho', 'lon_rho', -92.0, 'z_rho')
salt_slice.coords['lat_rho'] = xisoslice(ds, 'xi_rho', 'lon_rho', -92.0, 'lat_rho')
```
In this last case, the coordinates also need to be sliced in a similar way, hence the last two lines.

**Comment by jbusecke** [comment]
Dear @hetland, sorry for the long delay. I think this looks fantastic (will probably use this today in my science application 😁), but could you explain why this would be suited for xgcm and not xarray?
This looks to me like it only acts on one set of dimensions of several xarrays. This might be a candidate to integrate into xarray directly? Or am I missing something here? 

We are currently [exploring options](https://discourse.pangeo.io/t/plotting-ocean-variables-in-density-coordinates-using-xhistogram/125/12) to build a more general vertical regridding/remapping function as part of the [cmip6-hackathon](https://www.us-ocb.org/cmip6-hackathon-october-16-18-2019/). I would love to eventually have several options, which server different purposes: 
1. A slicing/indexing approach (find the nearest index, or interpolate variables onto another surface). Can you generalize your approach to use interpolation to find the values?
2. A binning approach, which conserves properties, simiar to @rabernat s approach above.

I think having both gives users the flexibility to achieve a bunch of goals.

**Comment by hetland** [comment]
Yes, @jbusecke, you are correct that this acts only on two xarray arrays. In fact, the dataset reference (ds) could be removed I think, and users could pass references to the arrays directly. I think I just wrote it that way since that's what my development code looked like.

I would be happy to have something like this in xarray, and it seems to be something that wouldn't clutter up the xarray codebase too much, since it's pretty short. There are no dependencies so it would be easy to maintain. One potential problem is that this approach only allows a projection, and the isosurface needs to be in an field that is monotonic in the projected dimension. This might cause confusion to users who want a generic isosurface code instead. Perhaps calling it `iso_interpolation` or `iso_projection` (or something similar) would make it more clear? Or perhaps it would just be more confusing... In any case, a good clear name would be critical.

Regarding the regridding, remapping, I have been following that discussion on discourse closely. To clarify, I think that using a histogram to convert between a fixed coordinate and a tracer coordinate (e.g., density) is a 'regridding', whereas the code above is more like an 'interpolation'. I think they are different enough that they should be considered as completely separate tasks, although they do both serve the general idea of getting out of grid space.

**Comment by jbusecke** [comment]
Fair point. I would suggest to raise a similar issue over at xarray. Once that is integrated either xgcm or a new 'super-duper-vertical-transformation-package' resulting from the hackathon could use this (xarray will be a dependencie anyways).

**Comment by jbusecke** [comment]
Just looking through some older issues and found this. I recently started developing some code to make regridding (finding a new depth grid, possibly defined on tracer surface in space and time) and remapping (converting data like tracer concentrations, velocities, etc from old to new depth grid) [here](https://github.com/jbusecke/xarrayutils/blob/master/doc/vertical_coords.ipynb).

This uses a bit different philosophy but also works as pure xarray. The conservative remapping is pretty fast but the linear interpolation not (since I am wrapping it in `vectorize` during the call to `apply_ufunc`), but I have started to work on implementing numba, which is blazing fast in my preliminary tests. 

I think this is the most general approach I have seen so far. E.g. one can use a pure interpolation for regridding + remapping, equivalent to what you have done here, but also choose different options. There is a lot of interest in these kinds of methods and I think it would be great if we could come up with a framework to include all needed cases. 

ccing @cspencerjones @raphaeldussin
